File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Users/Classes/UsersDataSource/UsersDataSource.m
//
//  UsersDataSource.m
//  LoginComponent
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "UsersDataSource.h"
#import "UserTableViewCell.h"
#import <Quickblox/Quickblox.h>
#import "Profile.h"
#import "PlaceholderGenerator.h"
#import "User.h"

@interface UsersDataSource()

@property (strong, nonatomic) NSMutableArray <QBUUser *> *_selectedUsers;
@property (strong, nonatomic) NSMutableArray <QBUUser *> *users;

@end

@implementation UsersDataSource

- (instancetype)init
{
    self = [super init];
    if (self) {
        self._selectedUsers = [NSMutableArray array];
        self.users = [NSMutableArray array];
    }
    return self;
}

#pragma mark - Public methods

- (void)updateUsers:(NSArray<QBUUser *> *)users {
    for (QBUUser *chatUser in users) {
        [self updateUser:chatUser];
    }
}

- (void)updateUser:(QBUUser *)user {
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"ID == %@", @(user.ID)];
    QBUUser *localUser = [[self.users filteredArrayUsingPredicate:predicate] firstObject];
    if (localUser) {
        //Update local User
        localUser.fullName = user.fullName;
        localUser.updatedAt = user.updatedAt;
        return;
    }
    [self.users addObject:user];
}

- (NSArray<QBUUser *> *)selectedUsers {
    
    return [self._selectedUsers copy];
}

- (void)selectUserAtIndexPath:(NSIndexPath *)indexPath {
    
    QBUUser *user = self.usersSortedByLastSeen[indexPath.row];
    
    if ([self._selectedUsers containsObject:user]) {
        [self._selectedUsers removeObject:user];
    }
    else {
        [self._selectedUsers addObject:user];
    }
}

- (QBUUser *)userWithID:(NSUInteger)ID {
    
    for (QBUUser *user in self.users) {
        
        if (user.ID == ID) {
            return user;
        }
    }
    
    return nil;
}

- (NSArray <NSNumber *> *)idsForUsers:(NSArray <QBUUser *>*)users {
    
    NSMutableArray *result = [NSMutableArray arrayWithCapacity:users.count];
    
    for (QBUUser *user in users) {
        [result addObject:@(user.ID)];
    }
    
    return result;
}

- (void)removeAllUsers {
    
    [self.users removeAllObjects];
}

- (NSArray <QBUUser *> *)usersSortedByFullName {
    
    return [self sortUsersBySEL:@selector(fullName)];
}

- (NSArray <QBUUser *> *)usersSortedByLastSeen {
    
    return [self sortUsersBySEL:@selector(updatedAt)];
}

- (NSArray <QBUUser *> *)sortUsersBySEL:(SEL)selector {

    Profile *profile = [[Profile alloc] init];
    User *me = [[User alloc] initWithID:profile.ID fullName:profile.fullName];
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"ID != %@", me.ID];
    NSArray *unsorterUsers = [self.users filteredArrayUsingPredicate:predicate];
    
    // Create sort Descriptor
    NSSortDescriptor *usersSortDescriptor =
    [[NSSortDescriptor alloc] initWithKey:NSStringFromSelector(selector)
                                ascending:NO];
    NSArray *sortedUsers = [unsorterUsers sortedArrayUsingDescriptors:@[usersSortDescriptor]];
    
    return sortedUsers;
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    return self.usersSortedByLastSeen.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    UserTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"UserCell"];
    
    QBUUser *user = self.usersSortedByLastSeen[indexPath.row];
    BOOL selected = [self._selectedUsers containsObject:user];
    NSString *name = user.fullName.length > 0 ? user.fullName : user.login;
    UIImage *userImage = [PlaceholderGenerator placeholderWithSize:CGSizeMake(32, 32)  title:name];
    
    [cell setFullName:name];
    [cell setCheck:selected];
    [cell setUserImage:userImage];
    
    return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    
    NSString *str = [NSString stringWithFormat:@"Select users for call (%tu)", self._selectedUsers.count];
    
    return NSLocalizedString(str, nil);
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Users/Classes/UsersDataSource/UsersDataSource.swift
//
//  UsersDataSource.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 12/10/18.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox

struct UsersDataSourceConstant {
    static let userCellIdentifier = "UserCell"
}

class UsersDataSource: NSObject {
    
    // MARK: - Properties
    var selectedUsers = [QBUUser]()
    private var users = [QBUUser]()
    
    func update(users: [QBUUser]) {
        for chatUser in users {
            update(user:chatUser)
        }
    }
    
    func update(user: QBUUser) {
        if let localUser = users.filter({ $0.id == user.id }).first {
            //Update local User
            localUser.fullName = user.fullName
            localUser.updatedAt = user.updatedAt
            return
        }
        users.append(user)
    }
    
    func selectUser(at indexPath: IndexPath) {
        
        let user = usersSortedByLastSeen()[indexPath.row]
        if selectedUsers.contains(user) {
            selectedUsers.removeAll(where: { element in element == user })
        } else {
            selectedUsers.append(user)
        }
    }
    
    func user(withID ID: UInt) -> QBUUser? {
        return users.filter{ $0.id == ID }.first
    }
    
    
    func ids(forUsers users: [QBUUser]) -> [NSNumber] {
        
        var result = [NSNumber]()
        
        for user in users {
            result.append(NSNumber(value: user.id))
        }
        return result
    }
    
    func removeAllUsers() {
        users.removeAll()
    }
    
    func usersSortedByFullName() -> [QBUUser] {
        let sortedUsers = unsortedUsersWithoutMe().sorted(by: {
            guard let firstUserName = $0.fullName, let secondUserName = $1.fullName else {
                return false
            }
            return firstUserName < secondUserName
        })
        return sortedUsers
    }
    
    func usersSortedByLastSeen() -> [QBUUser] {
        let sortedUsers = unsortedUsersWithoutMe().sorted(by: {
            guard let firstUpdatedAt = $0.updatedAt, let secondUpdatedAt = $1.updatedAt else {
                return false
            }
            return secondUpdatedAt < firstUpdatedAt
        })
        return sortedUsers
    }
    
    func unsortedUsersWithoutMe() -> [QBUUser] {
        var unsorterUsers = self.users
        let profile = Profile()
        if profile.isFull == false {
            return unsorterUsers
        }
        guard let index = unsorterUsers.index(where: { $0.id == profile.ID }) else {
            return unsorterUsers
        }
        unsorterUsers.remove(at: index)
        return unsorterUsers
    }
    
    //MARK: - Load User from server
    func loadUser(_ id: UInt, completion: ((QBUUser?) -> Void)? = nil) {
        QBRequest.user(withID: id, successBlock: { [weak self] (response, user) in
            self?.update(user: user)
            completion?(user)
        }) { (response) in
            completion?(nil)
        }
    }
}

extension UsersDataSource: UITableViewDataSource {
    // MARK: - UITableViewDataSource
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return usersSortedByLastSeen().count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        guard let cell = tableView.dequeueReusableCell(withIdentifier: UsersDataSourceConstant.userCellIdentifier)
            as? UserTableViewCell else {
                return UITableViewCell()
        }
        
        let user = usersSortedByLastSeen()[indexPath.row]
        let selected = selectedUsers.contains(user)
        
        let size = CGSize(width: 32.0, height: 32.0)
        var name = user.fullName ?? ""
        if name.isEmpty {
            name = user.login ?? "Unknown user"
        }
        let userImage = PlaceholderGenerator.placeholder(size: size, title: name)
        cell.fullName = name
        cell.check = selected
        cell.userImage = userImage
        
        return cell
    }

    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        
        let str = String(format: "Select users for call (%tu)", selectedUsers.count)
        return NSLocalizedString(str, comment: "")
    }
}

